File: POTD/Mathematical/Geometry/Cpp/3464. Maximize the Distance Between Points on a Square.cpp
// Time : O(log(side)*n*k*log(n))
// Space : O(2*n)
using ll = long long;
class Solution {
private:
    ll perimeter;
    bool check(int mid, vector<ll>& flattened, int k) {
        int n = flattened.size();
        for (int i = 0; i < n; i++) {
            ll startPos = flattened[i];
            int cnt = 1;

            ll lastPos = startPos;

            while (cnt < k) {
                ll target=lastPos+mid;
                auto it = lower_bound(flattened.begin() + i + 1,
                                      flattened.end(),target);

                if (it == flattened.end()) {
                    break;
                }

                cnt++;
                lastPos=*it;
            }

            if(cnt==k && (startPos+perimeter-lastPos)>=mid){
                return true;
            }
        }

        return false;
    }

public:
    int maxDistance(int side, vector<vector<int>>& points, int k) {

        perimeter = 1LL*4 * side;
        vector<ll> flattened;

        for (auto& p : points) {
            int x = p[0];
            int y = p[1];

            ll dist;
            if (y == 0) {
                dist = x;
            } else if (x == side) {
                dist = side + y;
            } else if (y == side) {
                dist = 1LL*3 * side - x;
            } else if (x == 0) {
                dist = 1LL*4 * side - y;
            }

            flattened.push_back(dist);
            flattened.push_back(dist + perimeter);
        }

        sort(begin(flattened), end(flattened));

        ll low = 0, high = 2 * side;
        ll res;
        while (low <= high) {
            ll mid = low + (high - low) / 2;


            if (check(mid, flattened, k)) {
                res = mid;
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }

        return res;
    }
};

File: POTD/Mathematical/Geometry/Python/3464. Maximize the Distance Between Points on a Square.py

class Solution:
    def maxDistance(self, side: int, points: List[List[int]], k: int) -> int:
        perimeter=4*side
        flattened=[]

        for i in range(len(points)):
            x=points[i][0]
            y=points[i][1]

            dist=-1
            if y==0:
                dist=x
            elif x==side:
                dist=side+y
            elif y==side:
                dist=3*side-x
            elif x==0:
                dist=4*side-y
            
            flattened.append(dist)
            flattened.append(dist+perimeter)
        
        flattened.sort()

        low=0
        high=2*side

        res=0
        while low<=high:
            mid=low+(high-low)//2

            if self.check(mid,flattened,k,perimeter):
                res=mid
                low=mid+1
            else:
                high=mid-1
        
        return res
    
    def check(self,mid:int,flattened:List[List[int]],k:int,perimeter:int):
        n=len(flattened)

        for i in range(n):
            startPos=flattened[i]
            cnt=1

            lastPos=startPos

            while cnt<k:
                target=lastPos+mid
                idx = bisect_left(flattened, target, i + 1)

                if idx>=n:
                    break
                cnt+=1
                lastPos=flattened[idx]
            
            if cnt==k and (startPos+perimeter-lastPos)>=mid:
                return True
        
        return False
